pkg/web_app/lib/src/widget/downloads_chart/widget.dart

```diff
@@ -6,34 +6,260 @@ import 'dart:convert';
+
+    final tickLabel = SVGTextElement();
+    tickLabel.setAttribute(
+        'class', 'downloads-chart-tick-label  downloads-chart-tick-label-y');
+    tickLabel.text =
+        '${compactFormat(i * interval).value}${compactFormat(i * interval).suffix}';
+    tickLabel.setAttribute('x', '${xMax + padding}');
+    tickLabel.setAttribute('y', '$y');
+    chart.append(tickLabel);
+
+    if (i == 0) {
+      // No long tick in the bottom, we have the x-axis here.
+      continue;
+    }
+
+    final longTick = SVGPathElement();
+    longTick.setAttribute('class', 'downloads-chart-frame');
+    longTick.setAttribute('d',
+        'M${xZero - (lineThickness / 2)} $y L${xMax - (lineThickness / 2)} $y');
+    chart.append(longTick);
+  }
+
+  // We use the clipPath to cut the ends of the chart lines so that we don't
+  // draw outside the frame of the chart.
+  final clipPath = SVGClipPathElement();
+  clipPath.setAttribute('id', 'clipRect');
+  final clipRect = SVGRectElement();
+  clipRect.setAttribute('y', '$yMax');
+  clipRect.setAttribute('height', '${chartheight - (lineThickness / 2)}');
+  clipRect.setAttribute('x', '$xZero');
+  clipRect.setAttribute('width', '$chartWidth');
+  clipPath.append(clipRect);
+  chart.append(clipPath);
+
+  // Chart lines and legends
+
+  final lines = <StringBuffer>[];
+  for (int versionRange = 0; versionRange < values[0].length; versionRange++) {
+    final line = StringBuffer();
+    var c = 'M';
+    for (int week = 0; week < values.length; week++) {
+      final (x, y) = computeCoordinates(
+          computeDateForWeekNumber(newestDate, values.length, week),
+          values[week][versionRange]);
+      line.write(' $c$x $y');
+      c = 'L';
+    }
+    lines.add(line);
+  }
+
+  double legendXCoor = xZero;
+  double legendYCoor =
+      tickLabelYCoordinate + firstTickLabel.getBBox().height + labelPadding;
+  final legendWidth = 20;
+  final legendHeight = 8;
+
+  for (int i = 0; i < lines.length; i++) {
+    final path = SVGPathElement();
+    path.setAttribute('class', '${strokeColorClass(i)} downloads-chart-line ');
+    // We assign colors in reverse order so that main colors are chosen first for
+    // the newest versions.
+    path.setAttribute('d', '${lines[lines.length - 1 - i]}');
+    path.setAttribute('clip-path', 'url(#clipRect)');
+    chart.append(path);
+
+    final legend = SVGRectElement();
+    chart.append(legend);
+    legend.setAttribute('class',
+        'downloads-chart-legend ${fillColorClass(i)} ${strokeColorClass(i)}');
+    legend.setAttribute('height', '$legendHeight');
+    legend.setAttribute('width', '$legendWidth');
+
+    final legendLabel = SVGTextElement();
+    chart.append(legendLabel);
+    legendLabel.setAttribute('class', 'downloads-chart-tick-label');
+    if (i == 5) {
+      // We have an 'other' line
+      legendLabel.text = 'Other';
+    } else {
+      legendLabel.text = ranges[ranges.length - 1 - i];
```

Comment:
Is there any chance that `ranges[5]` could contain the text `Other`, so you don't need to special case here?

Reply:
Done. Did a small refactor. Much better, thanks.